huaweicloud/services/ims/resource_huaweicloud_images_image_copy.go

```diff
@@ -0,0 +1,330 @@
+			"visibility": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"data_origin": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"disk_format": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"image_size": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"checksum": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"status": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceImsImageCopyCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+
+	var jobId string
+	imsClient, err := cfg.ImageV1Client(common.GetRegion(d, cfg))
+	if err != nil {
+		return diag.Errorf("error creating IMS client: %s", err)
+	}
+
+	targetRegion := d.Get("target_region").(string)
+	sourceRegion := cfg.GetRegion(d)
+	if targetRegion == "" || targetRegion == sourceRegion {
+		withinRegionCopyOpts := imagecopy.WithinRegionCopyOpts{
+			Name:                d.Get("name").(string),
+			Description:         d.Get("description").(string),
+			CmkId:               d.Get("kms_key_id").(string),
+			EnterpriseProjectID: common.GetEnterpriseProjectID(d, cfg),
+		}
+
+		log.Printf("[DEBUG] Within region copy Options: %#v", withinRegionCopyOpts)
+
+		sourceImageId := d.Get("source_image_id").(string)
+		jobRes, err := imagecopy.WithinRegionCopy(imsClient, sourceImageId, withinRegionCopyOpts).ExtractJobStatus()
+		if err != nil {
+			return diag.Errorf("error creating image copy within region: %s", err)
+		}
+		jobId = jobRes.JobID
+	} else {
+		crossRegionCopyOpts := imagecopy.CrossRegionCopyOpts{
+			Name:              d.Get("name").(string),
+			Description:       d.Get("description").(string),
+			TargetRegion:      targetRegion,
+			TargetProjectName: targetRegion,
+			AgencyName:        d.Get("agency_name").(string),
+			VaultId:           d.Get("vault_id").(string),
+		}
+
+		log.Printf("[DEBUG] Cross region copy Options: %#v", crossRegionCopyOpts)
+
+		sourceImageId := d.Get("source_image_id").(string)
+		jobRes, err := imagecopy.CrossRegionCopy(imsClient, sourceImageId, crossRegionCopyOpts).ExtractJobStatus()
+		if err != nil {
+			return diag.Errorf("error creating image copy cross region: %s", err)
+		}
+		jobId = jobRes.JobID
+	}
+
+	// Wait for the copy image to become available.
+	log.Printf("[DEBUG] Waiting for IMS to become available")
+	err = cloudimages.WaitForJobSuccess(imsClient, int(d.Timeout(schema.TimeoutCreate)/time.Second), jobId)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	imageId, err := cloudimages.GetJobEntity(imsClient, jobId, "image_id")
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId(imageId.(string))
+
+	// set tags
+	tagRaw := d.Get("tags").(map[string]interface{})
+	if len(tagRaw) > 0 {
+		// if the image is copied to new region, a new client needs to be created in the new region
+		imsClient, err = getImsV2Client(d, cfg)
```

Comment:
I think we should call this function before copying image, because it may be fail in cross-region by password authentication type.

Reply:
done